### ISSUE TITLE
yolo v8 - Object Tracking problem - tracker_id=None 

### ISSUE THREAD

**Issue by rafaelgildin** [ultralytics/ultralytics#8922]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Predict

### Bug

When setting `match_thresh: 0.3` in `bytetrack.yaml` the tracker_id since the second frame becomes None. 
With `match_thresh: 0.5` that doesn't happen, but a lot of the people in the frame aren't detected, that's why I reduced the tracker parameters of confidence.  I can't show the images for security reasons, but can try with a public dataset if needed. 

Related issues:
- https://github.com/ultralytics/ultralytics/issues/1739 - not useful since it skips the None problem. 

Possible solutions I thought:
- implement the logic to keep the history of track_ids and if tracker_id is None, get the track_id based on the distance between the current bounding box detected and the past not None boundind box.
- implement the tracker like `https://www.youtube.com/watch?v=QCG8QMhga9k`, but I think the result can be similar since it's the same processing being done in yolo tracking.
- Train the detection model with custom data, looking to improve the model confidence, hence working with `match_thresh: 0.5`. But again, this will skip the current problem.

### Environment

`yolo checks`
Ultralytics YOLOv8.1.27 🚀 Python-3.10.12 torch-2.2.1+cu121 CUDA:0 (NVIDIA GeForce RTX 4070 Laptop GPU, 8188MiB)
Setup complete ✅ (20 CPUs, 31.2 GB RAM, 93.2/250.9 GB disk)

OS                  Linux-5.15.146.1-microsoft-standard-WSL2-x86_64-with-glibc2.35
Environment         Linux
Python              3.10.12
Install             pip
RAM                 31.19 GB
CPU                 13th Gen Intel Core(TM) i9-13900H
CUDA                12.1

matplotlib          ✅ 3.8.3>=3.3.0
opencv-python       ✅ 4.9.0.80>=4.6.0
pillow              ✅ 10.2.0>=7.1.2
pyyaml              ✅ 6.0.1>=5.3.1
requests            ✅ 2.31.0>=2.23.0
scipy               ✅ 1.12.0>=1.4.1
torch               ✅ 2.2.1>=1.8.0
torchvision         ✅ 0.17.1>=0.9.0
tqdm                ✅ 4.66.2>=4.64.0
psutil              ✅ 5.9.8
py-cpuinfo          ✅ 9.0.0
thop                ✅ 0.1.1-2209072238>=0.1.1
pandas              ✅ 2.2.1>=1.1.4
seaborn             ✅ 0.13.2>=0.11.0

### Minimal Reproducible Example

- bytetrack.yaml
```
tracker_type: bytetrack
track_high_thresh: 0.1
track_low_thresh: 0.1 
new_track_thresh: 0.1
track_buffer: 1000 
match_thresh: 0.3 
```

- python code
```
from ultralytics import YOLO
from PIL import Image
import cv2 as cv
import supervision as sv
from collections import defaultdict
import numpy as np
import os

def change_to_project_path():
    project_path = os.getcwd()
    if 'notebooks' in project_path: 
        project_path = project_path.replace('/notebooks','')
        os.chdir(f'{project_path}')

project_path = change_to_project_path()
DEBUG = 1

model_path = os.path.join('models','yolov8n.pt')
model = YOLO(model_path)
video_path = os.path.join('data','videos','video_file.mp4')
output_path_not_formatted = video_path.replace('.mp4', f'_{task}_out_nf.mp4')
output_path = video_path.replace('.mp4', f'_{task}_out.mp4')

# Get the video codec and properties from the input video 
cap = cv.VideoCapture(video_path)
fps = int(cap.get(cv.CAP_PROP_FPS))
width = int(cap.get(cv.CAP_PROP_FRAME_WIDTH))
height = int(cap.get(cv.CAP_PROP_FRAME_HEIGHT))
frame_count = int(cap.get(cv.CAP_PROP_FRAME_COUNT))
total_duration = frame_count / fps 
print(f"Video (sec): {0}-{round(total_duration,2)}")
print(f"Frames: {0}-{frame_count}")

# Write output video
fourcc = cv.VideoWriter_fourcc(*'mp4v')
out = cv.VideoWriter(output_path_not_formatted, fourcc, fps, (width, height))

i = 0
while cap.isOpened():
    success, frame = cap.read()
    if success:
        result = model.track(frame, verbose=False, persist=True, conf=0.1, iou=0.1, tracker='bytetrack.yaml')[0]

        # Visualize the results on the frame
        bounding_box_annotator = sv.BoundingBoxAnnotator()
        label_annotator = sv.LabelAnnotator()
        detections = sv.Detections.from_ultralytics(result)
        detections = detections[detections.class_id == 0]

        labels = []        
        for _,_,prob,class_id,tracker_id,_ in detections:
            s = f"#{tracker_id} {round(prob*100)}%"
            labels.append(s)

        annotated_image = bounding_box_annotator.annotate(scene=frame.copy(), detections=detections)
        annotated_image = label_annotator.annotate(scene=annotated_image, detections=detections, labels=labels)
        
        if DEBUG:
            sv.plot_image(image=annotated_image, size=(16, 16))
            if(i == 10): break        
        out.write(annotated_image)
        i += 1
    else:
        break
cap.release()
out.release()
```

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by rafaelgildin** [comment]
@glenn-jocher 

1. IOU is actually very low to avoid the large amount of overlaping boxes and conf too, just for testing purposes, then to look if it’s a tracking or detection problem. Apparently is a tracking problem so I could train a custom dataset to detect people.
2. I did increased the track_buffer to 1000 frames, which covers all frames from the test video. 
3. I can set the verbose to True, but it increases the inference time and the debugging time, but will do later when deployed.
4. Totally agree, but it takes time to label and train, but I’m doing it.
5. It could maintain the track continuity, but I should be careful to not duplicate the track_id. Thanks for the example implementation, but I should calculate the current number of bounding_boxes and align with the historical situation where this number appears. Then calculate the track_id based on distance and treat the duplicates too.
6. Does it make sense to implement the tracking separately from the object detection since I’m going to change its logic from point 5 or is it better to stay with yolo tracking?

Thanks.

**Comment by glenn-jocher** [comment]
@rafaelgildin Thanks for the detailed feedback! 🙌 It sounds like you're making solid progress in troubleshooting your tracking challenge.

1. Given your situation, customizing your dataset for better detection of people seems like a sound approach. 🧐
2. Great to hear you've already adjusted the `track_buffer`! 
3. Understandable regarding `verbose=True` potentially slowing things down—optimizing for deployment is key.
4. Labeling and training do take time, but it's an investment that typically pays off in accuracy. Keep at it!
5. Your plan for ensuring track continuity sounds logical. Ensuring unique `track_id`s while considering historical context is indeed important for maintaining the integrity of tracking.
6. Regarding implementing tracking separately: It might offer you more flexibility to experiment and fine-tune the logic specifically for your needs. If you're considering significantly altering the built-in tracking logic, running it as a separate post-detection process could indeed make sense. This way, you can maintain the benefits of YOLO for detection while customizing tracking to fit your exact scenario.

It’s always a balance between leveraging existing efficient implementations and customizing to fit specific scenarios. Let us know how it goes and if you run into more questions. Happy coding!

**Comment by rafaelgildin** [comment]
@glenn-jocher 
Thanks for the points and feedbacks. I mostly agree with all.

**Comment by glenn-jocher** [comment]
@rafaelgildin You're welcome! I'm glad the feedback was helpful 😊. If you decide to go down the route of implementing custom logic for your tracking, don't hesitate to share your journey or ask for further guidance. Your exploration can offer valuable insights to the community. Happy experimenting, and looking forward to hearing about your progress!

**Comment by rafaelgildin** [comment]
Thanks @glenn-jocher ❤️

**Comment by glenn-jocher** [comment]
@rafaelgildin you're welcome! If any more questions come up or you need further assistance as you customize your logic, feel free to reach out. Happy coding! 💻🚀

**Comment by akdenmikail** [comment]
> @rafaelgildin you're welcome! If any more questions come up or you need further assistance as you customize your logic, feel free to reach out. Happy coding! 💻🚀
hellooo ,Now think about it, there are 20 people in an area. Here, I detect 20 people using yolov8 like this results = model(frame,persist=True, classes=classes_to_count). But when I use our model for tracking, yolov8 and results = model.track(frame,persist=True, classes=classes_to_count), 5 out of 20 people are detected and tracked. Why does the track part of the yolov8 model track so few of my 20 people?

Can you help me?

**Comment by Y-T-G** [comment]
@akdenmikail You just download that yaml file, and then change the confidence value to a lower one. Then you just pass it as argument.

`results = model.track(ASSETS / "bus.jpg", tracker="bytetrack.yaml") `

Pass the path to the yaml file you modified.

**Comment by Y-T-G** [comment]
@akdenmikail Why do you want to change `register_track`? I don't understand.

**Comment by akdenmikail** [comment]
> @akdenmikail Why do you want to change `register_track`? I don't understand.

Hello, I tried this method, here the person detection is the same, 10 people out of 20, but the id keeps changing during the tracking.

**Comment by akdenmikail** [comment]
> @rafaelgildin you're welcome! If any more questions come up or you need further assistance as you customize your logic, feel free to reach out. Happy coding! 💻🚀

hello ,How are you? I am getting an error regarding the track.

**Comment by Y-T-G** [comment]
@akdenmikail By default, `ultralytics` uses botsort.yaml which you can also download from the URL in the `trackers` folder and change the confidence.

To keep the track ID consistent, you need to pass `persist=True`.

**Comment by Y-T-G** [comment]
You just need to pass `tracker=bytetrack.yaml`. The rest of the arguments you're using are the same.

**Comment by akdenmikail** [comment]
results = model.track(frame,persist=True,tracker="bytetrack.yaml", classes=classes_to_count)
I ran it like this but it was not enough to track. Thank you for your help. I would like to communicate with you better. Can you give me an address to reach me?

**Comment by Y-T-G** [comment]
@akdenmikail You can join the Ultralytics Discord.

https://discord.com/invite/n6cFeSPZdD

**Comment by akdenmikail** [comment]
> @akdenmikail You can join the Ultralytics Discord.
> 
> https://discord.com/invite/n6cFeSPZdD

Discord has been closed in our country. I apologize. Is there another platform I can communicate with?

**Comment by Y-T-G** [comment]
@akdenmikail What's the issue currently? Did you download the yaml file and modify `new_track_thresh` to 0.25?

Change the yaml file name to `custom.yaml` and then use `tracker="custom.yaml"`

**Comment by akdenmikail** [comment]
> @akdenmikail What's the issue currently? Did you download the yaml file and modify `new_track_thresh` to 0.25?
> 
> Change the yaml file name to `custom.yaml` and then use `tracker="custom.yaml"`

The current problem is that when I use this yaml, the id of the person who starts to be followed changes and gives me the wrong count in the field.

**Comment by Y-T-G** [comment]
@akdenmikail Download this yaml, change `new_track_thresh` to 0.25 and rename the yaml file to `custom.yaml` and then use `tracker="custom.yaml"`. Delete the old yaml that you downloaded.

https://github.com/ultralytics/ultralytics/blob/main/ultralytics/cfg/trackers/botsort.yaml

**Comment by glenn-jocher** [comment]
I'm glad to help here! For tracking issues, ensure your `bytetrack.yaml` settings are optimized for your scenario. For further assistance, please continue using this platform.

**Comment by akdenmikail** [comment]
> @akdenmikail Download this yaml, change `new_track_thresh` to 0.25 and rename the yaml file to `custom.yaml` and then use `tracker="custom.yaml"`. Delete the old yaml that you downloaded.
> 
> https://github.com/ultralytics/ultralytics/blob/main/ultralytics/cfg/trackers/botsort.yaml

Okay, I'll try it and share the results with you. Thank you very much.

**Comment by glenn-jocher** [comment]
I'm glad to help here! For tracking issues, ensure your `bytetrack.yaml` settings are optimized for your scenario. For further assistance, please continue the discussion here on GitHub.

**Comment by glenn-jocher** [comment]
@akdenmikail i'm glad to help here! For tracking issues, ensure your `bytetrack.yaml` settings are optimized for your scenario. For further assistance, please continue the discussion here.

**Comment by akdenmikail** [comment]
> @akdenmikailBurada yardımcı olmaktan mutluluk duyarım! İzleme sorunları için, ayarlarınızın `bytetrack.yaml`senaryonuz için optimize edildiğinden emin olun. Daha fazla yardım için lütfen tartışmaya burada devam edin.

https://github.com/ultralytics/ultralytics/pull/17172#issuecomment-2446239560

**Comment by glenn-jocher** [comment]
I'm glad to help here on GitHub! For tracking issues, ensure your `bytetrack.yaml` settings are optimized for your scenario. Unfortunately, we can't provide private support, but feel free to continue the discussion here.
